recipes/Fine_Tuning_Granite/Finetuning_Granite.ipynb

```diff
@@ -0,0 +1,505 @@
+        "start_time = timeit.default_timer()\n",
+        "# Start training\n",
+        "trainer.train()\n",
+        "training_time = timeit.default_timer() - start_time\n"
+      ],
+      "metadata": {
+        "id": "oZ2rUKO2sdae"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## Saving the Fine-tuned Model  \n",
+        "  \n",
+        "After the training process is complete, it's crucial to save our fine-tuned model. This step ensures that we can reuse the model later without having to retrain it. We'll save both the model weights and the tokenizer, as they work in tandem to process and generate text.  \n",
+        "  \n",
+        "Saving the model allows us to distribute it, use it in different environments, or continue fine-tuning it in the future. It's a critical step in the machine learning workflow, preserving the knowledge our model has acquired through the training process.  "
+      ],
+      "metadata": {
+        "id": "zUjBCOZVIaas"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "model.save_pretrained(\"./results\")\n",
+        "tokenizer.save_pretrained(\"./results\")"
+      ],
+      "metadata": {
+        "id": "rK1i3CFeebzu"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "### Persisting the Model to Huggingface Hub  \n",
+        "  \n",
+        "After fine-tuning and validating our model, a optional step is to make it easily accessible for future use or sharing with the community. The Hugging Face Hub provides an excellent platform for this purpose.  \n",
+        "  \n",
+        "Uploading our model to the Hugging Face Hub offers several benefits:  \n",
+        "1. Easy sharing and collaboration with other researchers or developers  \n",
+        "2. Version control for your model iterations  \n",
+        "3. Integration with various libraries and tools in the Hugging Face ecosystem  \n",
+        "4. Simplified deployment options  \n",
+        "  \n",
+        "We'll demonstrate how to push our fine-tuned model and tokenizer to the Hugging Face Hub, making it available for others to use or for easy integration into other projects. This step is essential for reproducibility and for contributing to the broader NLP community.  "
+      ],
+      "metadata": {
+        "id": "se2YObHlj4f3"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "from google.colab import userdata\n",
+        "\n",
+        "model.push_to_hub(\"rawkintrevo/granite-3b-code-instruct-2k-legal\",\n",
+        "                  token= userdata.get('HF_TOKEN'))"
+      ],
+      "metadata": {
+        "id": "JkL-yHtth3Kc"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## Loading the Fine-tuned Model  \n",
```

Comment:
Would it make sense to move this section to another recipe and link this one to it? Since we have the model in-hand already.

Reply:
Yea... This was for adam who was worried about giving the workshop live and it taking an hour+ to train the model. I put this little backdoor in so he could get everyone to the point where training had started, then tell them, "OK, now restart your session, we'll just load the finetuned model and play with it now".

I'm open to other ways to skin the cat.

Reply:
I see. Makes sense.